### PR TITLE
feat: add heap profile collection

### PR DESCRIPTION
**Pull request by nolanmar511**

### CODE REVIEW

File: ts/src/profilers/heap-profiler.ts [hunk 1]
```diff
@@ -1,7 +1,58 @@
...
+import {perftools} from '../../../proto/profile';
+import {serializeHeapProfile} from './profile-serializer';
+
+const profiler = require('bindings')('sampling_heap_profiler');
+
 export class HeapProfiler {
-  constructor() {
-    throw new Error('Heap Profiler is unimplemented');
+  private enabled: boolean;
+  /**
```
Comment (aalexand) [inline ts/src/profilers/heap-profiler.ts]: Nit: add an empty line before this one.

File: ts/src/profilers/heap-profiler.ts [hunk 2]
```diff
@@ -1,7 +1,58 @@
...
+   */
+  constructor(private intervalBytes: number, private stackDepth: number) {
+    this.enable();
+  }
+
+  /**
+   * Collects a heap profile when heapProfiler is enabled. Otherwise throws
+   * an error.
+   *
+   * @param durationMillis - recorded duration of the profile.
```
Comment (aalexand) [inline ts/src/profilers/heap-profiler.ts]: Heap profiles do not have duration, they are instantaneous.

File: ts/src/profilers/heap-profiler.ts [hunk 3]
```diff
@@ -1,7 +1,58 @@
...
+   */
+  profile(durationMillis: number): perftools.profiles.IProfile {
+    if (!this.enabled) {
+      throw new Error('Heap profiler not enabled.');
```
Comment (aalexand) [inline ts/src/profilers/heap-profiler.ts]: Nit: "... is not enabled." to make the message a proper sentence.

File: ts/src/profilers/heap-profiler.ts [hunk 4]
```diff
@@ -1,7 +1,58 @@
...
+    }
+    const result = profiler.getAllocationProfile();
+    const startTimeNanos = Date.now() * 1000 * 1000;
+    return serializeHeapProfile(
+        result, startTimeNanos, durationMillis * 1000 * 1000,
```
Comment (aalexand) [inline ts/src/profilers/heap-profiler.ts]: See above on duration.

Comment (nolanmar511) [reply ts/src/profilers/heap-profiler.ts]: Done.

File: ts/src/profilers/heap-profiler.ts (hunk 3 above)
Comment (nolanmar511) [reply ts/src/profilers/heap-profiler.ts]: Done.

File: ts/src/profilers/heap-profiler.ts (hunk 2 above)
Comment (nolanmar511) [reply ts/src/profilers/heap-profiler.ts]: Done.

File: ts/src/profilers/heap-profiler.ts (hunk 1 above)
Comment (nolanmar511) [reply ts/src/profilers/heap-profiler.ts]: Done.

File: ts/src/profilers/heap-profiler.ts [hunk 5]
```diff
@@ -1,7 +1,55 @@
...
+    this.enable();
+  }
+
+  /**
+   * Collects a heap profile when heapProfiler is enabled. Otherwise throws
+   * an error.
+   */
+  profile(): perftools.profiles.IProfile {
+    if (!this.enabled) {
+      throw new Error('Heap profiler is not enabled.');
```
Comment (aalexand) [inline ts/src/profilers/heap-profiler.ts]: Do we catch such errors at the top level and make sure they just get logged into the profiler log and do not kill the customer's program?

Note that it's possible that customer starts their program with heap enabled on one machine and with heap disabled on another machine and the current behavior of the server is that the set of profile types is the same among all clients of the same deployment so the one with heap profiling disabled could be asked by the server to collect the heap data and the agent should ignore that logging a warning.

Comment (nolanmar511) [reply ts/src/profilers/heap-profiler.ts]: Yes -- these errors are caught at the top level and logged.

**Comment by aalexand** [comment]
@ofrobots Any comments on the code?

File: ts/test/test-heap-profiler.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    after(() => {
+      v8HeapProfiler.startSamplingHeapProfiler.restore();
+      v8HeapProfiler.stopSamplingHeapProfiler.restore();
+      v8HeapProfiler.getAllocationProfile.restore();
+      (Date.now as any).restore();
+    });
+
+    it('should return a profile equal to the expected profile', async () => {
+      const durationMillis = 10 * 1000;
```
Comment (ofrobots) [inline ts/test/test-heap-profiler.ts]: Unused?

File: ts/test/test-heap-profiler.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+      const intervalBytes = 1024 * 512;
+      const stackDepth = 32;
+      let profiler = new HeapProfiler(intervalBytes, stackDepth);
+      let profile = profiler.profile();
+      assert.deepEqual(heapProfile, profile);
+    });
+
+    it('should throw error when disabled', async () => {
+      const durationMillis = 10 * 1000;
```
Comment (ofrobots) [inline ts/test/test-heap-profiler.ts]: unused? If gts didn't catch this, can you open a bug against gts?

File: ts/test/test-heap-profiler.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      const intervalBytes = 1024 * 512;
+      const stackDepth = 32;
+      const profiler = new HeapProfiler(intervalBytes, stackDepth);
+      profiler.disable();
+      try {
```
Comment (ofrobots) [inline ts/test/test-heap-profiler.ts]: You can use [`assert.throws`](https://nodejs.org/dist/latest-v8.x/docs/api/assert.html#assert_assert_throws_block_error_message) for asserting that a snippet of code throws.

Comment (nolanmar511) [reply ts/test/test-heap-profiler.ts]: Does `assert.throws` work for asynchronous things?

I tried adding the following and the test failed.
```
assert.throws(
        async () => {
          throw new Error('ERROR');
        },
        Error
      );
```

But this passed:
```
assert.throws(
         () => {
          throw new Error('ERROR');
        },
        Error
      );
```

Comment (nolanmar511) [reply ts/test/test-heap-profiler.ts]: Never mind -- I forget that profiling isn't asynchronous for heap-profiler.

**Pull request merged**